### PR TITLE
Improve DictionaryBlock copy methods

### PR DESCRIPTION
**Pull request by pettyjamesm**
Improves `DictionaryBlock#copyRegion` and `#copyPositions` methods to detect scenarios that can avoid unnecessary work by removing the outer `DictionaryBlock` layer.

Also includes a one liner commit to perform the cheaper check first inside of `DictionaryBlock#getPositions`.

### CODE REVIEW

File: core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java [hunk 1]
```diff
@@ -418,7 +418,19 @@ public Block copyPositions(int[] positions, int offset, int length)
             }
             newIds[i] = newId;
         }
-        return new DictionaryBlock(dictionary.copyPositions(positionsToCopy.elements(), 0, positionsToCopy.size()), newIds);
+        Block compactDictionary = dictionary.copyPositions(positionsToCopy.elements(), 0, positionsToCopy.size());
+        if (positionsToCopy.size() == length) {
+            // discovered that all positions are unique, so return the unwrapped underlying dictionary directly
+            return compactDictionary;
+        }
+        return new DictionaryBlock(
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java:426]: please add tests to `io.trino.block.TestDictionaryBlock`

File: core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java [hunk 2]
```diff
@@ -437,6 +449,17 @@ public Block getRegion(int positionOffset, int length)
     public Block copyRegion(int position, int length)
     {
         checkValidRegion(positionCount, position, length);
+        // Avoid repeated volatile reads to the uniqueIds field
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java:452]: please add tests to `io.trino.block.TestDictionaryBlock`

File: core/trino-main/src/test/java/io/trino/block/AbstractTestBlock.java [hunk 3]
```diff
@@ -401,6 +402,16 @@ protected void assertPositionEquals(Block block, int position, Slice expectedByt
     {
     }
 
+    private static Block copyBlockToCompact(Block block)
+    {
+        if (block instanceof DictionaryBlock) {
+            return ((DictionaryBlock) block).compact();
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/test/java/io/trino/block/AbstractTestBlock.java]: Updated the method to just return the "compact size in bytes" and not imply that it returns a block that has been copied.

File: core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java [hunk 4]
```diff
@@ -394,9 +394,9 @@ public Block copyPositions(int[] positions, int offset, int length)
...
 
-        if (uniqueIds == positionCount) {
-            // each block position is unique, therefore it makes more sense to unwrap dictionary
-            // block by copying selected positions
+        if (length <= 1 || dictionary instanceof DictionaryBlock || uniqueIds == positionCount) {
+            // each block position is unique or the dictionary is a nested dictionary block,
+            // therefore it makes sense to unwrap this outer dictionary layer directly
             int[] positionsToCopy = new int[length];
             for (int i = 0; i < length; i++) {
                 positionsToCopy[i] = getId(positions[offset + i]);
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java:402]: Yep, that's why this version is only targeting scenarios where we know that either:
1. No id value will repeat from the selected positions (`length <= 1`, `uniqueIds == positionCount`)
2. Making the selected ids unique is going to be handled at the lower layer (`dictionary instanceof DictionaryBlock`)

File: core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java (hunk 1 above)
Comment (pettyjamesm) [reply core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java:426]: Added

File: core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java (hunk 2 above)
Comment (pettyjamesm) [reply core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java:452]: Added

**Comment by sopel39** [comment]
small comment

File: core/trino-main/src/test/java/io/trino/block/AbstractTestBlock.java [hunk 5]
```diff
@@ -401,6 +402,17 @@ protected void assertPositionEquals(Block block, int position, Slice expectedByt
     {
     }
 
+    private static long blockCompactSizeInBytes(Block block)
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/test/java/io/trino/block/AbstractTestBlock.java]: Done

**Pull request merged**